src/compose/ComposeBox.js

```diff
@@ -153,270 +136,352 @@ const updateTextInput = (textInput, text) => {
+      } else {
+        dispatch(sendTypingStart(narrow));
+      }
+      if (!isEditing) {
+        dispatch(draftUpdate(narrow, messageInputValue));
+      }
+    }
+  }, [props, messageInputState, prevMessageInputState]);
 
-  updateIsFocused = () => {
-    this.setState(state => ({
-      ...state,
-      isFocused: state.isMessageFocused || state.isTopicFocused,
-    }));
-  };
+  const updateIsFocused = useCallback(() => {
+    setFocusState(s => ({ ...s, either: s.message || s.topic }));
+  }, [setFocusState]);
 
-  getCanSelectTopic = () => {
-    const { isEditing, narrow } = this.props;
+  const getCanSelectTopic = useCallback(() => {
+    const { isEditing, narrow } = props;
     if (isEditing) {
       return isStreamOrTopicNarrow(narrow);
     }
     if (!isStreamNarrow(narrow)) {
       return false;
     }
-    return this.state.isFocused;
-  };
-
-  setMessageInputValue = (message: string) => {
-    updateTextInput(this.messageInputRef.current, message);
-    this.handleMessageChange(message);
-  };
-
-  setTopicInputValue = (topic: string) => {
-    updateTextInput(this.topicInputRef.current, topic);
-    this.handleTopicChange(topic);
-  };
-
-  insertMessageTextAtCursorPosition = (text: string) => {
-    const { message, selection } = this.state;
+    return focusState.either;
+  }, [props, focusState.either]);
 
-    this.setMessageInputValue(
-      message.slice(0, selection.start) + text + message.slice(selection.end, message.length),
-    );
-  };
+  const handleMessageChange = useCallback(
+    (newValue: string) => {
+      setMessageInputState(s => ({ ...s, value: newValue }));
+      setIsMenuExpanded(false);
+    },
+    [setMessageInputState, setIsMenuExpanded],
+  );
+
+  const setMessageInputValue = useCallback(
+    (updater: (typeof messageInputState) => string) => {
+      updateTextInput(messageInputRef.current, updater(messageInputState));
+      setMessageInputState(s => ({ ...s, value: updater(s) }));
+      setIsMenuExpanded(false);
+    },
+    [messageInputState, setIsMenuExpanded],
+  );
 
-  insertVideoCallLinkAtCursorPosition = (url: string) => {
-    const { _ } = this.props;
-    const linkMessage = _('Click to join video call');
-    const linkText = `[${linkMessage}](${url})`;
+  const handleTopicChange = useCallback(
+    (newValue: string) => {
+      setTopicInputState({ value: newValue });
+      setIsMenuExpanded(false);
+    },
+    [setTopicInputState, setIsMenuExpanded],
+  );
 
-    this.insertMessageTextAtCursorPosition(linkText);
-  };
+  const setTopicInputValue = useCallback(
+    (topic: string) => {
+      updateTextInput(topicInputRef.current, topic);
+      handleTopicChange(topic);
+    },
+    [handleTopicChange],
+  );
+
+  const insertMessageTextAtCursorPosition = useCallback(
+    (text: string) => {
+      setMessageInputValue(
+        s =>
+          s.value.slice(0, s.selection.start)
+          + text
+          + s.value.slice(s.selection.end, s.value.length),
```

Comment:
As we discussed in the office:
* It's silly that this clobbers any selected text (between `s.selection.start` and `s.selection.end`). That's unlikely to ever be helpful. Inserting at `s.selection.end`, without clobbering, would be a clear improvement.
* Arguably inserting at the cursor at all is unhelpful -- it could be in the middle of some markup like a link, for example -- and it should go at the end. Though there are quite plausible use cases where the cursor is preferable, e.g. you're composing a long message and want to insert some files/images in the middle between two paragraphs.
  * I briefly thought that making that change, so that this didn't need to look at the selection state, would help with the `setMessageInputState` vs `updateTextInput` issue above. But then I think we found that the issue is still there and fundamentally the same even with the selection state out of the picture.